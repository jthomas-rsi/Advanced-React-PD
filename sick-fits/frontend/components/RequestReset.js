import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import DisplayError from './ErrorMessage';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

const RequestReset = () => {
  // form controls and state
  const { inputs, handleChange, resetForm, clearForm } = useForm({
    email: '',
  });

  // mutation hook
  const [signup, { data, loading, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  // handle the form submission
  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await signup().catch((error) => console.error(error));
    console.log(res);
    console.log({ data, loading, error });
    resetForm();
    // send the email and password to the graphqlAPI
  };

  return (
    <Form method="POST" onSubmit={(e) => handleSubmit(e)}>
      <h2>Request A Password Reset</h2>
      <DisplayError error={error} />
      <fieldset>
        {data?.sendUserPasswordResetLink === null && (
          <p>Success - Check your email for a link to reset your password. </p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="text"
            id="email"
            name="email"
            placeholder="Your email address"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <button type="submit" onClick={() => {}}>
          Reset Password
        </button>
      </fieldset>
    </Form>
  );
};

export default RequestReset;
