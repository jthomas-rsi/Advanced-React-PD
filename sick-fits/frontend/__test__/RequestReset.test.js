import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset';

const email = 'james@isAwesome.com';
const mocks = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

const TestComponent = () => (
  <MockedProvider mocks={mocks}>
    <RequestReset />
  </MockedProvider>
);

describe('RequestReset Testing suite ', () => {
  it('Component renders and matches a snapshot', () => {
    const { container } = render(TestComponent());
    expect(container).toMatchSnapshot();
  });
  it('Component calls mutation when rendered ', async () => {
    render(TestComponent());
    userEvent.type(screen.getByPlaceholderText(/your email address/i), email);
    userEvent.click(screen.getByRole('button', { name: /reset password/i }));
    const successMessage = await screen.findByText(/success/i);
    expect(successMessage).toBeInTheDocument();
  });
});
