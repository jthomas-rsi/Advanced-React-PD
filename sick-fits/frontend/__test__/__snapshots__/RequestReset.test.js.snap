// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestReset Testing suite  Component renders and matches a snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 jnelna"
    method="POST"
  >
    <h2>
      Request A Password Reset
    </h2>
    <fieldset>
      <label
        for="email"
      >
        Email
        <input
          id="email"
          name="email"
          placeholder="Your email address"
          type="text"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Reset Password
      </button>
    </fieldset>
  </form>
</div>
`;
